vowpalwabbit/array_parameters.h

```diff
@@ -0,0 +1,172 @@
+#pragma once
+#include <stdio.h>
+#include <iostream>
+#include <stdlib.h>
+#include <string.h>
+
+#ifndef _WIN32
+#include <sys/mman.h>
+#endif
+
+using namespace std;
+typedef float weight;
+
+class weight_vector;
+
+class weights_iterator_iterator
+{
+private:
+	weight* _cur;
+public:
+	weights_iterator_iterator(weight* cur)
+		: _cur(cur)
+	{ }
+	
+	weight& operator*() { return *_cur; }
+
+	weights_iterator_iterator& operator++()
+	{
+		_cur++;
+		return *this;
+	}
+
+	template<typename T>
+	weights_iterator_iterator operator+(T index) { return weights_iterator_iterator(_cur + index); }
+
+	template<typename T>
+	weights_iterator_iterator& operator+=(T index)
+	{
+		_cur += index;
+		return *this;
+	}
+
+	weights_iterator_iterator& operator=(const weights_iterator_iterator& other)
+	{
+		_cur = other._cur;
+		return *this;
+	}
+
+	bool operator==(const weights_iterator_iterator& rhs) { return _cur == rhs._cur; }
+	bool operator!=(const weights_iterator_iterator& rhs) { return _cur != rhs._cur; }
+
+};
+class weights_iterator
+{
+private:
+	weight* _current;
+	uint32_t _stride;
+
+public:
+	typedef weights_iterator_iterator w_iter;
+
+	weights_iterator(weight* current, uint32_t stride)
+		: _current(current), _stride(stride)
+	{ }
+
+	weight& operator*() { return *_current; }
+
+	weights_iterator& operator++()
+	{
+		_current += _stride;
+		return *this;
+	}
+
+	template<typename T>
+	weights_iterator operator+(T index) { return weights_iterator(_current + (index*_stride), _stride); }
+
+	template<typename T>
+	weights_iterator& operator+=(T index)
+	{
+		_current += (index*_stride);
+		return *this;
+	}
+
+	weights_iterator& operator=(const weights_iterator& other)
+	{
+		_current = other._current;
+		_stride = other._stride;
+		return *this;
+	}
+
+	bool operator==(const weights_iterator& rhs) { return _current == rhs._current; }
+	bool operator!=(const weights_iterator& rhs) { return _current != rhs._current; }
+
+	//to iterate within a bucket
+	w_iter begin() { return w_iter(_current); }
+	w_iter end(size_t offset) { return w_iter(_current + offset); }
+
+};
+
+class weight_vector //different name? (previously array_parameters)
+{
+private:
+	weight* _begin;
+	uint64_t _weight_mask;  // (stride*(1 << num_bits) -1)
+	uint32_t _stride_shift;
+	bool _seeded;
+
+public:
+	typedef weights_iterator iterator;
+
+	weight_vector()
+		: _begin(nullptr), _weight_mask((uint64_t)LONG_MAX), _stride_shift(0),_seeded(false)
```

Comment:
It is best to avoid having default constructors like this which don't really do anything.  But, if you feel you need it, you may want to consider an alternative: In C++ 11, you can also just set the member variables in the class, e.g.
bool _seeded = false;

Then you won't even need this.

Second, the correct C++ way of saying `_weight_mask((uint64_t)LONG_MAX)` is `_weight_mask(std::numeric_limits<type>::max())`.